server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -48,7 +48,7 @@ public DataStream(String name, String timeStampField, List<Index> indices, long
         this.indices = indices;
         this.generation = generation;
         assert indices.size() > 0;
-        assert indices.get(indices.size() - 1).getName().equals(getBackingIndexName(name, generation));
+        assert indices.get(indices.size() - 1).getName().endsWith(getBackingIndexName(name, generation));
```

Comment:
We can definitely relax the logic on the naming of backing indices, but I think the write index needs to stay the same. In other words, ILM may trigger a rollover which creates a new write index, but the current write index should always conform to the naming scheme.

Reply:
I started with that invariant in place but if we want to enforce that we will have to invalidate policies that don't have a `hot` phase (ie. a `rollover` action).
A policy that only has a `warm` phase (the scenario in the integration test https://github.com/elastic/elasticsearch/pull/57616/files#diff-fd8498322f0015d57152688358ddaa0aR45 ) will attempt to shrink the write index.
I believe we can fail in this particular case and not allow any index replacement (shrink, restore index) on the write index but then the user will have to manually add another index to the data stream (ie. changing the write index) in order for shrink to eventually be successful and the policy to be able to continue.

Reply:
We decided to not allow the shrink operation on the write index. This means the naming of the write index will remain consistent with the agreed naming scheme for backing indices.